File: design-patterns/01- Creational Patterns/04- Singleton/Python/singleton.py
# Singleton Logger class that will be responsible for logging different actions.
# The most popular way of creating a singleton in Python is using the metaclass approach:
# A metaclass is the class of a class. Like a class defines how an instance of the class behaves, a metaclass defines how a class behaves. A class is an instance of a metaclass.

class Singleton(type):

    def _init_(cls, name, bases, attrs, **kwargs):
        super()._init_(name, bases, attrs)
        cls._instance = None

    def _call_(cls, *args, **kwargs):
        if cls._instance is None:
            cls.instance = super().call_(*args, **kwargs)
        return cls._instance


class Logger(metaclass=Singleton):
    #some class
    pass

l1 = Logger()
l2 = Logger()
print(l1 is l2)

File: design-patterns/01- Creational Patterns/04- Singleton/C++/singleton.cpp
// Singleton Logger class that will be responsible for logging different actions.
// We have declared a static instance which prevents re-initialization of the variable. 
// The getInstance() method is used to access the static instance. This way we ensure that only one instance of the class is created.
// This is just a simple example. Production code should include a destructor to release resources, provide thread safety, etc.

#include <iostream>
using namespace std;

class Logger {
  private:
    static Logger* instance;
    // Private constructor so it can not be called
    Logger(){};
  public:
    // Static access method
    static Logger* getInstance() {
      if (instance == NULL) {
        instance = new Logger();
      }
      return instance;
    }
    void log(string str){
      cout << "Log: " << str << endl;
    }
};

// Instance initialized on demand
Logger* Logger::instance = NULL;

int main() {
    
    Logger* L = Logger::getInstance();
    
    L->log("Login");
    L->log("Like");
    L->log("Comment");
    L->log("Logout");
    
    return 0;
}